<a href="https://colab.research.google.com/github/juma-paul/deep_learning/blob/main/MNIST_MultiClassificatioin_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Multiclass Classication**

Next, you will train the two layer neural network to do multiclass classication on digits dataset. Digits data is similar to MNIST but has even
smaller pixel

In [1]:
# Import required packages
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf

In [2]:
# Load Dataset
from sklearn.datasets import load_digits
digits = load_digits(return_X_y=True, as_frame=True)

In [3]:
# Assign features to X and targets to y
X = digits[0]
y = digits[1]

In [4]:
# Features
X

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0           0.0        0.0        5.0       13.0        9.0        1.0   
1           0.0        0.0        0.0       12.0       13.0        5.0   
2           0.0        0.0        0.0        4.0       15.0       12.0   
3           0.0        0.0        7.0       15.0       13.0        1.0   
4           0.0        0.0        0.0        1.0       11.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        4.0       10.0       13.0        6.0   
1793        0.0        0.0        6.0       16.0       13.0       11.0   
1794        0.0        0.0        1.0       11.0       15.0        1.0   
1795        0.0        0.0        2.0       10.0        7.0        0.0   
1796        0.0        0.0       10.0       14.0        8.0        1.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
0           0.0        0.0        0.0        0.0  ...        0.0        0.0   
1           0.0        0.0        0.0        0.0  ...        0.0        0.0   
2           0.0        0.0        0.0        0.0  ...        5.0        0.0   
3           0.0        0.0        0.0        8.0  ...        9.0        0.0   
4           0.0        0.0        0.0        0.0  ...        0.0        0.0   
...         ...        ...        ...        ...  ...        ...        ...   
1792        0.0        0.0        0.0        1.0  ...        4.0        0.0   
1793        1.0        0.0        0.0        0.0  ...        1.0        0.0   
1794        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1795        0.0        0.0        0.0        0.0  ...        2.0        0.0   
1796        0.0        0.0        0.0        2.0  ...        8.0        0.0   

      pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
0           0.0        0.0        6.0       13.0       10.0        0.0   
1           0.0        0.0        0.0       11.0       16.0       10.0   
2           0.0        0.0        0.0        3.0       11.0       16.0   
3           0.0        0.0        7.0       13.0       13.0        9.0   
4           0.0        0.0        0.0        2.0       16.0        4.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        2.0       14.0       15.0        9.0   
1793        0.0        0.0        6.0       16.0       14.0        6.0   
1794        0.0        0.0        2.0        9.0       13.0        6.0   
1795        0.0        0.0        5.0       12.0       16.0       12.0   
1796        0.0        1.0        8.0       12.0       14.0       12.0   

      pixel_7_6  pixel_7_7  
0           0.0        0.0  
1           0.0        0.0  
2           9.0        0.0  
3           0.0        0.0  
4           0.0        0.0  
...         ...        ...  
1792        0.0        0.0  
1793        0.0        0.0  
1794        0.0        0.0  
1795        0.0        0.0  
1796        1.0        0.0  

[1797 rows x 64 columns]

In [5]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1797 entries, 0 to 1796
Data columns (total 64 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pixel_0_0  1797 non-null   float64
 1   pixel_0_1  1797 non-null   float64
 2   pixel_0_2  1797 non-null   float64
 3   pixel_0_3  1797 non-null   float64
 4   pixel_0_4  1797 non-null   float64
 5   pixel_0_5  1797 non-null   float64
 6   pixel_0_6  1797 non-null   float64
 7   pixel_0_7  1797 non-null   float64
 8   pixel_1_0  1797 non-null   float64
 9   pixel_1_1  1797 non-null   float64
 10  pixel_1_2  1797 non-null   float64
 11  pixel_1_3  1797 non-null   float64
 12  pixel_1_4  1797 non-null   float64
 13  pixel_1_5  1797 non-null   float64
 14  pixel_1_6  1797 non-null   float64
 15  pixel_1_7  1797 non-null   float64
 16  pixel_2_0  1797 non-null   float64
 17  pixel_2_1  1797 non-null   float64
 18  pixel_2_2  1797 non-null   float64
 19  pixel_2_3  1797 non-null   float64
 20  pixel_2_

In [6]:
# Targets
y

0       0
1       1
2       2
3       3
4       4
       ..
1792    9
1793    0
1794    8
1795    9
1796    8
Name: target, Length: 1797, dtype: int64

In [7]:
y.info()

<class 'pandas.core.series.Series'>
RangeIndex: 1797 entries, 0 to 1796
Series name: target
Non-Null Count  Dtype
--------------  -----
1797 non-null   int64
dtypes: int64(1)
memory usage: 14.2 KB


In [8]:
# Convert the target to categorical variable
y_encoded = tf.keras.utils.to_categorical(y, num_classes=10)

In [9]:
# split the data into training and test sets
X_train_unscaled, X_test_unscaled, y_train, y_test = train_test_split(X, y_encoded, test_size=0.3, random_state=42)

In [10]:
# Scale the training data
scaler=StandardScaler()

X_train=scaler.fit_transform(X_train_unscaled)
X_test = scaler.transform(X_test_unscaled)

In [11]:
# Convert data into tensors
X_train = tf.convert_to_tensor(X_train, dtype = tf.float32)
X_test = tf.convert_to_tensor(X_test, dtype = tf.float32)
y_train = tf.convert_to_tensor(y_train, dtype = tf.float32)
y_test= tf.convert_to_tensor(y_test, dtype = tf.float32)

In [12]:
X_train.shape, X_test.shape

(TensorShape([1257, 64]), TensorShape([540, 64]))

**Question 8: Fill in the input dimension and output dimension of the two layer neural network appropriate for this dataset. (5 pts)**

In [13]:
# Set dimensions of the neural network
input_dim = X_train.shape[1]
hidden_dim = input_dim * 3
output_dim = 10

# Validate dimensions
assert input_dim > 0
assert hidden_dim > input_dim
assert output_dim == 10

print(f"Network Architecture:\n")
print(f"Input Layer: {input_dim} nodes")
print(f"Hidden Layer: {hidden_dim} nodes")
print(f"Output Layer: {output_dim} node")

Network Architecture:

Input Layer: 64 nodes
Hidden Layer: 192 nodes
Output Layer: 10 node


**Question 9: Define tensorflow variables for weights W1, b1, W2, and b2. Then, initialize both biases b1 and b2 to be 0 and initialize W1 andW2 by picking values uniformly at random from the interval [0, 0.1]. (5 pts)**

In [14]:
# Initialize weights for hidden layer (Layer 1)
W1 = tf.Variable(
    tf.random.uniform(
        shape=[input_dim, hidden_dim],
        minval=0,
        maxval=0.1,
        dtype=tf.float32,
        name='W1'
    )
)

# Initialize bias for hidden layer
b1 = tf.Variable(
    tf.zeros(
        shape=[hidden_dim],
        dtype=tf.float32,
        name='b1'
    )
)

# Initialize weights for output layer (Layer 2)
W2 = tf.Variable(
    tf.random.uniform(
        shape=[hidden_dim, output_dim],
        minval=0,
        maxval=0.1,
        dtype=tf.float32,
        name='W2'
    )
)

# Initialize bias for output layer
b2 = tf.Variable(
    tf.zeros(
        shape=[output_dim],
        dtype=tf.float32,
        name='b2'
    )
)

# Print parameter shapes for verification
print(f"Network Parameter Shapes:\n")
print(f"W1: {W1.shape} - Hidden layer weights")
print(f"b1: {b1.shape} - Hidden layer bias")
print(f"W2: {W2.shape} - Output layer weights")
print(f"b2: {b2.shape} - Output layer bias")

Network Parameter Shapes:

W1: (64, 192) - Hidden layer weights
b1: (192,) - Hidden layer bias
W2: (192, 10) - Output layer weights
b2: (10,) - Output layer bias


**Question 10: Complete the classier function** below to define a two layer neural network that outputs class probabilities for each class. Here,inputs is a matrix of shape n*d,where ith row of the inputs matrix contains xTi . (10 pts)

You may use tf.softmax.nn() function to compute softmax class probabilities.

In [15]:
def classifier(inputs: tf.Tensor) -> tf.Tensor:

    hidden = tf.matmul(inputs, W1) + b1
    hidden_activated = tf.nn.relu(hidden)

    logits = tf.matmul(hidden_activated, W2) + b2
    probabilities = tf.nn.softmax(logits)

    return probabilities

**Question 11: Complete the function below to compute cross entropy loss given one hot encoding of targets and predictions with classprobabilities for each prediction. Write the function such that autodiff will work, and note that you cannot use inbuilt tensorow cross entropyfunction. (10 pts)**

In [16]:
def cross_entropy(predictions, targets):
    epsilon = 1e-15 # prevents numerical instability by avoiding log(0) which would result
                    # in infinite/NaN values that break gradient calculations
    cr_entropies = -tf.reduce_sum(targets * tf.math.log(predictions + epsilon), axis=1)
    mean_cr_entropy = tf.reduce_mean(cr_entropies)

    return mean_cr_entropy

**Question 12: Complete the function below to train the neural network classifier using gradient descent. (5 pts)**

In [17]:
def train(inputs: tf.Tensor, targets: tf.Tensor, optimizer: tf.keras.optimizers.Optimizer) -> tf.Tensor:

    with tf.GradientTape() as tape:
        # Ensure variables are tracked
        tape.watch([W1, b1, W2, b2])

        # Forward pass
        predictions = classifier(inputs)
        loss = cross_entropy(predictions, targets)

    # Compute gradients
    trainable_vars = [W1, b1, W2, b2]
    gradients = tape.gradient(loss, trainable_vars)

    # Update variables using gradient descent
    optimizer.apply_gradients(zip(gradients, trainable_vars))


    return loss

**Train the model using 500 GD iterations.**

In [18]:
# Set hyper parameters
num_iterations = 200
learning_rate = 0.01
optimizer = tf.keras.optimizers.Adam(learning_rate)

for step in range(num_iterations):
    loss = train(X_train, y_train, optimizer)

    if (step + 1) % 20 == 0:
        print(f"Loss at step {step + 1}: {loss:.4f}")

Loss at step 20: 0.0703
Loss at step 40: 0.0038
Loss at step 60: 0.0011
Loss at step 80: 0.0007
Loss at step 100: 0.0005
Loss at step 120: 0.0004
Loss at step 140: 0.0004
Loss at step 160: 0.0003
Loss at step 180: 0.0003
Loss at step 200: 0.0002


Suppose, given the vector of class probabilities, you output the label with the highest class probability as your label. As an evaluation of ourmodel, we want to compute the number of mistakes that the model makes. For instance, if is the true label and is the prediction of the model, we will evaluate our model on this point with 0-1 loss 1( ≠ y) = {1 yˆ≠ y and 0 yˆ=y}
Over points, we will compute the mean 0-1 loss,1/n∑(i=1..n) 1 *(yˆi≠yi)

**Question 13: Compute the mean 0-1 loss of your classif[link text](https://)er on the test data. (10 pts)**

In [19]:
def compute_zero_one_loss(X_test: tf.Tensor, y_test: tf.Tensor) -> float:

    # Get predictions
    predictions = classifier(X_test)

    # Convert to class labels
    pred_labels = tf.argmax(predictions, axis=1)
    true_labels = tf.argmax(y_test, axis=1)

    # Compute 0-1 loss (1 if prediction != true label, 0 otherwise)
    mistakes = tf.cast(pred_labels != true_labels, tf.float32)

    # Return mean 0-1 loss
    return float(tf.reduce_mean(mistakes))

mean_loss = compute_zero_one_loss(X_test, y_test)
print(f"Mean 0-1 Loss: {mean_loss:.4f}")

Mean 0-1 Loss: 0.0185


The classifier achieves a **mean 0-1 loss** of **0.0185**, translating to an outstanding accuracy of approximately **98.15%**.